CorsixTH/Lua/dialogs/staff_dialog.lua

```diff
@@ -314,12 +314,7 @@ end
 
 function UIStaff:placeStaff()
   self.staff.pickup = true
-  self.staff:setNextAction({
-    name = "pickup",
-    ui = self.ui,
-    todo_close = self,
-    must_happen = true,
-  }, true)
+  self.staff:setNextAction(PickupAction():setUi(self.ui):setTodoClose(self):setMustHappen(), true)
```

Comment:
The aim of this patch was to fix a lot of formatting nuisances I had for a long time. For the actions, it was mostly lack of parentheses, but the big literals didn't look nicely either, and a few experiments with a more function-call like syntax proved to be very useful.

I haven't studied the actions in much detail, and I am not sure I want to go there. At the surface it all looks very much duplicated and overly complicated to me. I believe the reason is that the code is trying to decide the next action from the end of the current action. (Would love some feedback about that, but it's not trivial to say the least.)

Some actions (probably the older code, like toilets, and gp) hook that into the "on-exit" call, leading to nested persistent recursive functions. Other actions push things into the action_queue (probably was added after deciding nested persistent recursive functions aren't very nice). Then you get problems like pick_up, or interrupts like getting thirsty or going to the toilets, which break such a queue, leading to more work-arounds. Someone messes up and you get an empty action queue, which then leads to new code to decide what to do in that case.

So, in all, I believe the core problem is the chain of actions. Instead, I think there should be a point in a humanoid or object(?) that tracks what it is aiming to do long and short term, and generate actions from that. An action is then not much more than an animation. All logic of what to do next moves to that tracking point.
One form of such a point is a state machine, but I am not sure it's the best solution.

One thing I haven't decided what to make of it, are the registered callbacks that happen all over the place. I think they complicate matters, but don't know how badly.

As for your questions, technically the ", true" is the replacement priority for already existing actions rather than the priority of the new action. (I agree however you can see it as the priority of the new action as well.)
I haven't looked at the details of that priority flag, but it looks like very bad smelling code to me. In fact, the entire action queue smells, as I just explained. Currently, it looks like a waste of time trying to fix actions, to me, as explained above.